pkg/scheduler/core/util.go

```diff
@@ -68,22 +72,46 @@ func calAvailableReplicas(clusters []*clusterv1alpha1.Cluster, spec *workv1alpha
 		return availableTargetClusters
 	}
 
-	// Get the minimum value of MaxAvailableReplicas in terms of all estimators.
+	// get estimators and sort it by priority.
 	estimators := estimatorclient.GetReplicaEstimators()
+	sort.Slice(estimators, func(i, j int) bool {
+		return estimators[i].GetPriority() > estimators[j].GetPriority()
+	})
+
+	priorityOfLastAvailableEstimator := estimatorclient.EstimatorPriority(math.MinInt32)
 	ctx := context.WithValue(context.TODO(), util.ContextKeyObject,
 		fmt.Sprintf("kind=%s, name=%s/%s", spec.Resource.Kind, spec.Resource.Namespace, spec.Resource.Name))
+
 	for _, estimator := range estimators {
+		// if higher-priority estimators have formed a full result of member clusters, no longer to call lower-priority estimator.
+		if estimator.GetPriority() < priorityOfLastAvailableEstimator && len(clusterIndexToEstimatorPriority) == len(clusters) {
```

Comment:
This is the part where I feel complicated.
If we can make sure the estimator is looped by priority, then each estimator can focus on clusters that haven't been estimated.

Reply:
@Garrybest suggest grouping the estimators with different priorities, e.g.,[{estimators:[es1, es3], priority:9}, {estimators:[es2, es4], priority:8}, ...]

@whitewindmills suggest using `github.com/emirpasic/gods/trees/redblacktree`

@RainbowMango suggest looping by priority

So, combine your three suggestions, how about using `github.com/emirpasic/gods/trees/redblacktree`, the key is `priority`, the value is `map[string]estimators{}` with same priority, and loop by `priority` key?

Reply:
actually, `github.com/emirpasic/gods/maps/treemap` is more suitable.

Reply:
Hey guys, I think we have a misunderstanding here. It's not about rbtree or slice, both are OK for me. Let me give a example:

```
	// sort by priority
	sort.Slice(estimators, func(i, j int) bool {
		return estimators[i].GetPriority() > estimators[j].GetPriority()
	})

	ctx := context.WithValue(context.TODO(), util.ContextKeyObject,
		fmt.Sprintf("kind=%s, name=%s/%s", spec.Resource.Kind, spec.Resource.Namespace, spec.Resource.Name))

	// group by priority
	var estimatorsByPriority [][]estimatorclient.ReplicaEstimator
	for i, j := 0, 0; i < len(estimators); {
		for ; j < len(estimators); j++ {
			if estimator[j].GetPriority() != estimators[i].GetPriority() {
				break
			}
		}
		estimatorsByPriority = append(estimatorsByPriority, estimators[i:j])
		i = j
	}

loop:
	for _, estimatorGroup := range estimatorsByPriority {
		for _, estimator := range estimatorGroup {
			res, err := estimator.MaxAvailableReplicas(ctx, clusters, spec.ReplicaRequirements)
			if err != nil {
				klog.Errorf("Max cluster available replicas error: %v", err)
				continue
			}
			for i := range res {
				if res[i].Replicas == estimatorclient.UnauthenticReplica {
					continue
				}
				if availableTargetClusters[i].Name == res[i].Name && availableTargetClusters[i].Replicas > res[i].Replicas {
					availableTargetClusters[i].Replicas = res[i].Replicas
				}
			}
			// estimator with higher priority finished, get the result and skip other ones
			break loop
		}
	}
```

Just group them and add one more loop. It is better?

Reply:
I actually understand it that way, but I've implemented it using a treemap to make code more clean.

The reason I'm torn between using a treemap instead of a redblackmap is that the redblackmap is not convenient to get values by key.